[a11y] CupertinoSwitch On/Off labels

gilnobrega opened the pull request:
### Description

Displays I/O labels on CupertinoSwitch when `onOffSwitchLabels` accessibility property is enabled.

Adds `onOffSwitchLabel` property to `MediaQuery` and related classes. This is similar to other accessibility properties such as `highContrast`.

#### Off state
![image](https://github.com/flutter/flutter/assets/82336674/7e5a874b-b1a2-4429-a435-005e3a0a4dd3)

#### On state
![image](https://github.com/flutter/flutter/assets/82336674/7e294e60-6041-4bb6-a892-d7ea56cfa2c8)

#### Animation
![ezgif com-video-to-gif](https://github.com/flutter/flutter/assets/82336674/9803d256-b058-4a7c-bbaa-2b0253508885)

Addresses feedback of closed PR #39993. (that `onOffSwitchLabels` needs to come from `MediaQuery`)
Fixes issue #4830.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

gilnobrega committed:
CupertinoSwitch on off labels

gilnobrega committed:
Implements MediaQuery onOffSwitchLabels

gilnobrega committed:
implements directionality on labels

gilnobrega committed:
applies reaction opacity to labels

gilnobrega committed:
allows overriding on off label colors

gilnobrega committed:
rename constants

gilnobrega committed:
fix MediaQuery.onOffSwitchLabelsOf

gilnobrega committed:
add goldens

gilnobrega committed:
add mediaquery onOffSwitchLabel tests

gilnobrega committed:
fix onOffSwitchLabelsOf test

gilnobrega committed:
fixes label opacity
fixes off label color

gilnobrega committed:
Merge branch 'master' into cupertinoswitch-accessibility-labels

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -467,6 +491,12 @@ const double _kTrackInnerEnd = _kTrackWidth - _kTrackInnerStart;
+const double _kOnLabelPaddingHorizontal = 22.0 / 2.0;
+const double _kOffLabelWidth = 2.0 / 2.0;
+const double _kOffLabelPaddingHorizontal = 24.0 / 2.0;
+const double _kOffLabelRadius = 10.0 / 2.0;
```
Is there a reason why these are written as a function instead of just the value? As in, `1.0` instead of `2.0 / 2.0`.

gilnobrega replied:
Yes, this is because `22.0`, `2.0`, `24.0` and `10.0` are the actual size in the inspector,
see https://github.com/flutter/flutter/issues/4830#issuecomment-528495360

I'm happy to change these values to `11.0`, `1.0`, etc. though, should I do that?

MitchellGoodwin replied:
I think so. It will be a bit more clear. Adding a comment on where those values came from could be helpful as well.

MitchellGoodwin reviewed (commented):
Mostly looks good to me with one question.

gilnobrega committed:
comment for source of label sizes

vleeuwenmenno commented:
+1 really essential feature. Would be nice to have this merged soon.

MitchellGoodwin reviewed (approved):
LGTM! Thank you for putting this together.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -584,6 +622,36 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
     markNeedsPaint();
   }
 
+  bool get onOffSwitchLabels => _onOffSwitchLabels;
```
nit: this could have been `(Color, Color)?`, so the widget doesn't repaint if the label is off and the colors change.

justinmc commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -798,6 +904,102 @@ void main() {
+
+    await expectLater(
+      find.byKey(switchKey),
+      matchesGoldenFile('switch.tap.off.dark.labels.png'),
```
You can probably do all of these tests without using a golden, which would make them faster and more robust.  You can probably use the fact that `drawCircle` is used in one case and `drawRect` in another.  These can be tested with `paints`.  See for example [another PR that did something similar](
https://github.com/flutter/flutter/pull/127757/files#diff-4efbb5702e52581d71b67c66a97fa1cee641ad56bd541065a10f30ffb1efbc92R89) recently.

gilnobrega replied:
I could do that but aren't goldens better for the mid-animation effects? e.g.: `cupertino.switch.tap.turningOn.labels`,  `cupertino.switch.tap.turningOff.labels`, etc.

These test the switch labels opacity mid-animation.

justinmc replied:
Yeah I guess the mid-animation goldens are more reasonable.  Maybe keep those two?  If you think they're going to be useful for catching bugs or accidental changes to the visuals.

gilnobrega replied:
Replaced all new goldens with `PaintPattern`s.

The mid animation goldens were replaced with `PaintPattern`s with transparent colors.

justinmc reviewed (commented):
Thanks for finally tackling this old issue!  I think this looks fine except that the golden tests can probably be replaced with `paints`.  See my comment below.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -133,6 +135,16 @@ class CupertinoSwitch extends StatefulWidget {
+
+  /// The color to use for the accessibility label when the switch is off.
+  ///
+  /// Defaults to [Color(0xFFB3B3B3)] when null.
```
Why's the default on color a dynamic color while the off color is a hard coded one? Isn't the off color supposed to change in dark mode / high contrast mode etc.?

LongCatIsLooong replied:
Huh I asked the same question in https://github.com/flutter/flutter/pull/39993 and it seems the answer is no. But does `onLabelColor` respond to dark mode / high contrast?

gilnobrega replied:
Nice spot, I forgot to check for high contrast.

It seems that `onLabelColor` stays the same in high contrast or dark mode. However, the `offLabelColor` becomes white in high contrast in both light and dark modes.

This behaviour doesn't correspond to any of the `CupertinoColors` though.

Screenshots from Settings app, iOS 16.5.1:

##### Light mode
<img width="462" alt="lightmode" src="https://github.com/flutter/flutter/assets/82336674/5df5362f-7d4a-42ab-821d-4268cf406ac3">

##### Light mode, High contrast
<img width="462" alt="image" src="https://github.com/flutter/flutter/assets/82336674/a3af3c52-7158-40a0-bf28-226b44ded81d">

##### Dark mode
<img width="462" alt="image" src="https://github.com/flutter/flutter/assets/82336674/5e90b0c0-79a9-437a-909d-dc71f59bbe59">

##### Dark mode, High contrast
<img width="462" alt="image" src="https://github.com/flutter/flutter/assets/82336674/99423340-2832-45d1-91c2-a3b99f49dc26">

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -649,6 +717,56 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
+      final double leftLabelOpacity = visualPosition * (1.0 - currentReactionValue);
+      final double rightLabelOpacity = (1.0 - visualPosition) * (1.0 - currentReactionValue);
+
+      late final Offset onLabelOffset;
```
these don't have to be `late`?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -467,6 +491,14 @@ const double _kTrackInnerEnd = _kTrackWidth - _kTrackInnerStart;
 const double _kSwitchHeight = 39.0;
+// Label sizes and padding taken from xcode inspector.
+// See https://github.com/flutter/flutter/issues/4830#issuecomment-528495360
+const double _kOnLabelWidth = 1.0;
```
The measurements here are a bit different from https://github.com/flutter/flutter/pull/39993 (e.g. the stroke width being 1px instead of 2px in the previous PR)?

gilnobrega replied:
I explained why these values are half the value shown in the inspector in [this thread](https://github.com/flutter/flutter/pull/127776#discussion_r1231306321).

The width in #39993 is actually 1.25px, I don't know why! See: https://github.com/flutter/flutter/pull/39993/files#diff-c8dc74e1119fcfc41cb08a151f40285dc4cde603a5da5d66b53b1dd6b7ad5529R531

LongCatIsLooong replied:
Oh are the values in #39993 in device pixels instead of logic pixels (UIKit points)? I thought the values were taken from the view hierarchy inspector so they should be in points.

gilnobrega replied:
is 1.0px ok for the width then?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -649,6 +717,56 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
+      switch (textDirection) {
+        case TextDirection.ltr:
+          onLabelOffset =
+              trackRect.centerLeft.translate(_kOnLabelPaddingHorizontal, 0);
```
nit: if the parent RenderObject gives this switch tight constraints then the switch's actual dimensions could be very different from `_kSwitchWidth x _kSwitchHeight`.

LongCatIsLooong replied:
(this is somewhat of an edge case so feel free to keep it as is and open a new issue for it).

gilnobrega committed:
Merge branch 'master' into cupertinoswitch-accessibility-labels

gilnobrega committed:
addressed feedback from @LongCatIsLooong
introduced onOffLabelColors nullable record
added high contrast variant of offLabelColor

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -357,6 +370,22 @@ class _CupertinoSwitchState extends State<CupertinoSwitch> with TickerProviderSt
+                  widget.onLabelColor ?? CupertinoColors.white,
+                  context,
+                ),
+                CupertinoDynamicColor.resolve(
```
Could you create a new `CupertinoDynamicColor` for this instead of putting the logic for handling high contrast here, and update the documentation for the default offLabelColor?

gilnobrega replied:
I created a new `CupertinoDynamicColor` in `CupertinoColor`.

Is this ok, or should it not be in here `CupertinoColors` color as it's not reused in other widgets?

LongCatIsLooong replied:
Keeping it in `switch.dart` sgtm

gilnobrega replied:
done

LongCatIsLooong reviewed (commented):
The golden image changes and the code changes LGTM modulo high contrast handling and comments from other reviewers. Thanks for adding this!

gilnobrega committed:
Merge branch 'master' into cupertinoswitch-accessibility-labels

gilnobrega committed:
adds CupertinoColors.offSwitchLabel
handles off label color logic

gilnobrega committed:
addressed feedback from @justinmc
replaced goldens with PaintPatterns

gilnobrega committed:
moves off label color to switch.dart

justinmc reviewed (approved):
LGTM :+1: 

Thanks for the test changes!
